### DISCUSSION TITLE
Better use wrapper functions with local init instead of global init?

### DISCUSSION THREAD

**Discussion by KVolker** [FuPeiJiang/VD.ahk#21] [Show and tell]
Hi,
I stumbled over the issue that VD stopped working after some switching between desktops and I had to reload my whole script. And I couldn't find out why this happened. To be honest, it is quite cumbersome to find out why someone elses code does not work as expected, so I didn't really start to find out what was happening. But I fixed it for my case. I removed the "global" VD.init proposed in the documentation, and instead wrote wrappers for the VD functions I need (just case by case), like this example:

```
SwitchDesktopByNumber(aTargetDesktop){
    v := new VD()
    v.init()
    v.goToDesktopNum(aTargetDesktop)
    v := ""
}

SendActiveWindowToDesktop(aTargetDesktop){
    v := new VD()
    v.init()
    v.MoveWindowToDesktopNum("A", aTargetDesktop)
    v := ""
}

ToggleActiveWindowOnAllDesktops(){
    v := new VD()
    v.init()
    v.TogglePinWindow("A")
    v := ""
}
```
The v := "" should call the garbage collector on the object. For now, that works. 

@FuPeiJiang: Cool library, much work you did, appreciated, thanks for providing that to us!

CU
Volker

**Comment by KVolker** [comment]
Just another comment: often I write wrapper functions or objects for calls to libraries I incorporate into my own code. It is generally more future-safe, as I have to change code only in one place if I want to use another library or have to fix something etc. So in fact I had already those wrappers, but only one-liners, which I just had to extend a little bit.

**Comment by FuPeiJiang** [comment]
I know that if `explorer.exe` is restarted it won’t work anymore, I was planning to catch when `explorer.exe` is restarted, but surely there are other reasons that causes it to fail

This fix is op, it fixes Any other reason that might make it fail, by restarting it every time

adding a hook to catch when `explorer.exe` restarts slows down the computer, restarting `init()` on every action is computationally less expensive, but it isn’t faster, there should be no difference in speed if `init()` runs in 0 seconds

**Reply by KVolker** [reply]
If there is a delay it is not so much perceptible at all. Especially not in my case as in my application I use key hooks in a way that I have different functions if I press the key one, two, or three times, so anyway I added a 200ms delay to make it wait for a potential next key press. So time for initializing VD is no problem.

Tx again!

**Reply by KVolker** [reply]
Me again. I measured the ticks used for the initialization and destruction of the VD object. My function to measure this is:
```
DebugOut(aString){
   if (Global Debugging) {
      Out := A_TickCount " | "  aString "`n"
      OutputDebug, %Out%
   }
}
```

The function I profiled is like this
```
SwitchDesktopByNumber(aTargetDesktop){
    DebugOut("Before creation and init")
    v := new VD()
    v.init()
    DebugOut("After creation and init")
    v.goToDesktopNum(aTargetDesktop)
    DebugOut("Before destruction")
    v := ""
    DebugOut("After destruction")
}
```

The result is 
```
956392593 | Before creation and init
956392593 | After creation and init
956392843 | Before destruction
956392843 | After destruction
```

Including the measuring function, this seems to mean that both creation and destruction of VD take 0ms. VD.goToDesktopNum(x) takes 50ms.
